core/blockchain_pruner.go

```diff
@@ -0,0 +1,127 @@
+package core
+
+import (
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/harmony-one/harmony/core/rawdb"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/utils"
+)
+
+const (
+	logPruningProgressBatchCount = 10000
+	PreserveBlockAmount          = 262144 // Preserve 8 epochs of blocks (2^15) * 8
+)
+
+// InitialBlockPruning deletes all blocks from the blockchain backwards
+// from the current block preserving 8 epochs worth of blocks
+func (bc *BlockChain) InitialBlockPruning() {
+	blockNumberToDelete := bc.GetInitialPruningStartBlockNum()
+	if blockNumberToDelete == 0 {
+		utils.Logger().Info().Msgf("Initial pruning canceled - start number is 0\n")
+		return
+	}
+	utils.Logger().Info().Msgf("Pruning all blocks before block number %d\n", blockNumberToDelete)
+	numDeletedBlocksThisBatch := 0
+	highestBlockNumInBatch := blockNumberToDelete - 1
+	for {
+		blockNumberToDelete--
+		pruned := bc.PruneBlock(blockNumberToDelete)
+		finishedBatch := blockNumberToDelete%logPruningProgressBatchCount == 0
+		if pruned {
+			numDeletedBlocksThisBatch++
+			if numDeletedBlocksThisBatch == 1 { // track the first deleted block for compaction later
+				highestBlockNumInBatch = blockNumberToDelete
+			}
+		}
+		if finishedBatch {
+			utils.Logger().Info().Msgf("Pruned back to block number %d. %d blocks were pruned. Continuing\n", blockNumberToDelete, numDeletedBlocksThisBatch)
+			rawdb.WriteBlockPruningState(bc.ChainDb(), bc.ShardID(), blockNumberToDelete)
+			if numDeletedBlocksThisBatch > 0 { // if we deleted anything this batch, do compaction
+				rawdb.CompactPrunedBlocks(bc.ChainDb(), bc.ShardID(), blockNumberToDelete, highestBlockNumInBatch)
+			}
+			if blockNumberToDelete == 0 {
+				break
+			}
+			numDeletedBlocksThisBatch = 0
+		}
+	}
+
+	utils.Logger().Info().Msgf("Finished pruning blocks\n")
+}
+
+// PruneBlock deletes the block with a given block number. Returns bool for whether a block was deleted.
+func (bc *BlockChain) PruneBlock(blockNumber uint64) bool {
+	if blockNumber == 0 { // do not prune genesis block
```

Comment:
you can add a mutex for PruneBlock to avoid race conditions.

Reply:
PruneBlock is not a easy fix. Adding a lock will not solve everything.  One approach I recommend is to have 
1. one bool flag for each block
2. a block number before which all blocks are checked to be pruned.